platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,436 @@
+We will pass the full round info (metadata + events) to all listeners, and they can pull from it what they need.
+
+#### Roster Changes
+
+When the roster changes, the Hashgraph algorithm must be made aware. It needs roster information to be able to come
+to consensus. While several modules must respond to roster changes, it is the Hashgraph module that "owns" the roster.
+Each round has an associated roster -- the roster that was used to come to consensus on that round. While the
+Hashgraph owns the roster, it does not _determine_ the roster, this happens in the Execution layer.
+
+At runtime, there is a service in Execution that exposes to the world API for adding nodes to, or removing nodes from,
+the address book. At some point in time, Execution will decide the time is right to construct a new consensus roster,
+and pass that roster to the consensus system. **This must be deterministic**. Each node in the network must assign the
+same roster to the same rounds. However, Consensus and Execution run in completely different threads, and possibly,
+in different processes.
+
+With this architecture, we introduce a new concept into `EventCore`. `EventCore` has a `payload`, which is a `byte[]`
+that the Execution layer can use to carry whatever payload is meaningful to it. The consensus layer needs an analogous
+section of the `EventCore` into which messages particular to Consensus can be placed. This section will be called
+`repeated control_messages`. A `ControlMessage` is a protobuf message used to control, in some way, the consensus
+system.
+
+TODO: I'm seriously thinking of having "system_transactions" and "user_transactions" instead of "control_messages" and
+"payload".
+
+`ControlMessage` has a one-of, among which is `RosterChangeMessage`. This message will have the new roster, and be
+signed by the network as a whole. Each honest node will submit a `RosterChangeMessage`, and the Hashgraph module will
+recognize these messages. When it encounters the **first** such message for a given roster in a round that has come to
+consensus, the very next round will use that new activated roster.
+
+### Event Creator Module
+
+Every node in the network participating in consensus is permitted to create events to gossip to peers. These events
+are used both for transmitting user transactions, and as the basis of the hashgraph algorithm for "gossiping about
+gossip". Therefore, the Event Creator has two main responsibilities:
+
+1. Create events with "other parent(s)" so as to help the hashgraph progress consensus
+2. Fill events with transactions to be sent to the network through Gossip
+
+#### Creating Events
+
+The Event Creator is configured with a `maximum_event_creation_frequency`, measured in events/sec. This is a network
+wide setting. If any node creates events more rapidly than this setting, then the node will be reported to the Sheriff.
+An event is not necessarily created at this frequency, but will be created at no more than this frequency.
```

Comment:
There is an algorithm I designed (but never got a chance to write down) that provides a non-heuristic way of enforcing maximum creation rate in a way that is observable at the consensus layer. Heuristics for enforcing maximum event creation rates are going to be messy and hard to tune. Ping me if you'd like to set up some time to discuss.

Reply:
Maybe we can connect and discuss in more detail. Thanks!